### PR TITLE
Remove use of std::iterator

### PR DESCRIPTION
**Pull request by dschuff**
It's deprecated in C++17

### CODE REVIEW

**Review by tlively — APPROVED** [review]
I guess `pointer` should probably be `const Type*`, but if that's going to cause issues I can fix that separately.

**Review by kripken — APPROVED** [review]

**Comment by dschuff** [comment]
Using `const Type*` works. I also added another warning suppression for the LLVM code.

**Review by tlively — APPROVED** [review]

File: src/wasm-type.h [hunk 1]
```diff
@@ -263,11 +263,14 @@ class Type {
...
-  struct Iterator : std::iterator<std::random_access_iterator_tag,
-                                  Type,
-                                  long,
-                                  Type*,
-                                  const Type&> {
+  struct Iterator {
+    // Iterator traits
+    using iterator_category = std::random_access_iterator_tag;
+    using value_type = Type;
+    using difference_type = long;
```
Comment (dschuff) [inline src/wasm-type.h:270]: actually, should this just be `ptrdiff_t` instead of `long`?

Comment (tlively) [reply src/wasm-type.h:270]: yep, probably.

**Pull request merged**